pkg/validation/applicationinstallation_test.go

```diff
@@ -0,0 +1,205 @@
+		expectedError string
+	}{
+		{
+			name:          "Create ApplicationInstallation Success",
+			ai:            ai,
+			expectedError: "[]",
+		},
+		{
+			name: "Create ApplicationInstallation Failure - ApplicationDefinitation doesn't exist",
+			ai: &appkubermaticv1.ApplicationInstallation{
+				Spec: func() appkubermaticv1.ApplicationInstallationSpec {
+					spec := ai.Spec.DeepCopy()
+					spec.ApplicationRef.Name = invalidResource
+					return *spec
+				}(),
+			}, expectedError: `[spec.applicationRef.name: Not found: "invalid"]`,
+		},
+		{
+			name: "Create ApplicationInstallation Failure - Invalid Version",
+			ai: &appkubermaticv1.ApplicationInstallation{
+				Spec: func() appkubermaticv1.ApplicationInstallationSpec {
+					spec := ai.Spec.DeepCopy()
+					spec.ApplicationRef.Version = appkubermaticv1.Version{Version: *semverlib.MustParse("3.2.3")}
+					return *spec
+				}(),
+			}, expectedError: `[spec.applicationRef.version: Not found: 3.2.3]`,
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			err := ValidateApplicationInstallationSpec(context.Background(), fakeClient, testCase.ai.Spec)
+			if fmt.Sprint(err) != testCase.expectedError {
+				if testCase.expectedError == "[]" {
+					testCase.expectedError = "nil"
+				}
+				t.Fatalf("expected error to be %s but got %v", testCase.expectedError, err)
+			}
+		})
+	}
+}
+
+// TestValidateApplicationInstallationSpec tests the validation for ApplicationInstallation creation.
+func TestValidateApplicationInstallationUpdate(t *testing.T) {
+	ad := getApplicationDefinition(defaultAppName)
+	updatedAD := getApplicationDefinition("updated-app")
+	fakeClient := fakectrlruntimeclient.
+		NewClientBuilder().
+		WithScheme(testScheme).
+		WithObjects(ad, updatedAD).
+		Build()
+
+	ai := getApplicationInstallation(defaultAppName, defaultAppName, defaultAppVersion)
+
+	testCases := []struct {
+		name          string
+		ai            *appkubermaticv1.ApplicationInstallation
+		updatedAI     *appkubermaticv1.ApplicationInstallation
+		expectedError string
+	}{
+		{
+			name: "Update ApplicationInstallation Success",
+			ai:   ai,
+			updatedAI: &appkubermaticv1.ApplicationInstallation{
+				Spec: func() appkubermaticv1.ApplicationInstallationSpec {
+					spec := ai.Spec.DeepCopy()
+					spec.Namespace.Labels = map[string]string{"key": "value"}
+					return *spec
+				}(),
+			},
+			expectedError: "[]",
+		},
+		{
+			name: "Update ApplicationInstallation Failure - .Namespace.Name is immutable",
+			ai:   ai,
+			updatedAI: &appkubermaticv1.ApplicationInstallation{
+				Spec: func() appkubermaticv1.ApplicationInstallationSpec {
+					spec := ai.Spec.DeepCopy()
+					spec.Namespace.Name = invalidResource
+					return *spec
+				}(),
+			},
+			expectedError: `[spec.namespace.name: Invalid value: "default": field is immutable]`,
+		},
+		{
+			name: "Update ApplicationInstallation Failure - .ApplicationRef.Name is immutable",
```

Comment:
`ApplicationRef` on a whole is immutable. Let me add another test for that use case 👍

Reply:
I think to be production-ready, we have to allow users to update and application version because deleting applicationInstallation means uninstalling the application and losing data if the workload was stateful which is not acceptable for upgrading a version of an application.

However, IIRC, when we write a proposal we said in the first version we make it immutable the time to dig a little bit about how to make difference between an admin deleting an application version from applicationDefinition and the user referencing a version that does not exist, and maybe some other corner cases around the upgrade...

Reply:
I think to be production-ready, we have to allow users to update and application version because deleting applicationInstallation means uninstalling the application and losing data if the workload was stateful which is not acceptable for upgrading a version of an application.

However, IIRC, when we write a proposal we said in the first version we make it immutable the time to dig a little bit about how to make difference between an admin deleting an application version from applicationDefinition and the user referencing a version that does not exist, and maybe some other corner cases around the upgrade...

Reply:
I think upgrading the version would be a bit complicated. My suggestion would be to work on that in a separate PR. We can update these webhooks later based on how we want to approach upgrades.

Reply:
I am ok with that :)

Reply:
sounds good to me